### PR TITLE
Fix set router epoch on runtime invalidate.

### PR DESCRIPTION
**Pull request by zalokhan**
## Overview

Description: CorfuRuntime: fetchLayout fetches the latest layout from a random layout server and sets the client router epochs to the newer, higher epoch number in a loop.
However, if the first router in the loop throws a network exception the loop is aborted and the other router epochs remain unset.

Why should this be merged: To avoid wrong epoch exceptions and set the router epochs correctly. 

Related issue(s) (if applicable): #1044 

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **d1fc0f5**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #1046 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/1GwIiS1vSUnWFujzkXQyZGv4HkysIJhd)

**Comment by zalokhan** [comment]
@rogermichoud Added a test. 
Verification: The test fails on master.

File: test/src/test/java/org/corfudb/runtime/CorfuRuntimeTest.java [hunk 1]
```diff
@@ -169,8 +195,70 @@ public void doesNotAllowReadsAfterSealAndBeforeNewLayout() throws Exception {
...
+     * @throws Exception
+     */
+    @Test
+    public void testFailedRouterInFetchLayout() throws Exception {
+
+        final Map<String, TestClientRouter> routerMap = new ConcurrentHashMap<>();
+        final AtomicReference<String> failedNode = new AtomicReference<>();
+
+        CorfuRuntime.overrideGetRouterFunction = (corfuRuntime, endpoint) -> {
+            if (failedNode.get() != null && endpoint.equals(failedNode.get())) {
```
Comment (no2chem) [inline test/src/test/java/org/corfudb/runtime/CorfuRuntimeTest.java]: Can you just move this logic to AbstractViewTest::getRouterFunction?

File: test/src/test/java/org/corfudb/runtime/view/AbstractViewTest.java [hunk 2]
```diff
@@ -350,6 +350,16 @@ public String getEndpoint(int port) {
         return "test:" + port;
     }
 
+    /**
```
Comment (no2chem) [inline test/src/test/java/org/corfudb/runtime/view/AbstractViewTest.java]: If you move the logic as suggested in the comment above, this function is no longer necessary.

File: test/src/test/java/org/corfudb/runtime/CorfuRuntimeTest.java (hunk 1 above)
Comment (no2chem) [reply test/src/test/java/org/corfudb/runtime/CorfuRuntimeTest.java]: Ah, I see what is happening here. Is there a reason to throw these exceptions (it seems like they aren't picked up in your test?)

Comment (no2chem) [reply test/src/test/java/org/corfudb/runtime/CorfuRuntimeTest.java]: Never mind, I see the logic. It's an unfortunate side effect of the test router not being able to simulate a closed connection. Add that to the comments (that this is simulating a failed node), and I think we'll be okay.

Comment (zalokhan) [reply test/src/test/java/org/corfudb/runtime/CorfuRuntimeTest.java]: Firstly the test routers never throw Network Exceptions and I needed to test them specifically.
These exceptions are caught by the CorfuRuntime while updating the router epochs.

This is very specific to this test and I wonder if anyone else would ever want to throw this exception at this point in the code.

Comment (rogermichoud) [reply test/src/test/java/org/corfudb/runtime/CorfuRuntimeTest.java]: I introduced a "simulated network exception" at the test router level. Can you not leverage this? It is way less fancy than that, so not sure it serves your purpose. That is what this method does, throwing a network exception from the router.

The method is simulateDisconnectedEndpoint(), in TestClientRouter.

Comment (zalokhan) [reply test/src/test/java/org/corfudb/runtime/CorfuRuntimeTest.java]: Sounds good I'll add a dependency on #982 and consume your test addition.

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **0ec14b0**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #1046 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/h2stK4gS0KbA8jVXxGU2T6xDbXmo0sxo)

**Comment by zalokhan** [comment]
@rogermichoud @no2chem 
Sorry for realizing this so late.
I am not able to reproduce the error using the new `simulateEndpointDisconnect` feature in the TestClientRouter.
This is because the NettyClientRouter unlike the TestClientRouter calls the `start()` method in the constructor. This at times can throw a NetworkException if the connection cannot be made.
Simulating this will require a different approach.

So for now, I'll stick to my existing test and make a few changes which @no2chem suggested in the review.

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **c51489f**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #1046 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/E60Hza0lil80uR2hvqNVZfMi71AP6cWX)

**Comment by no2chem** [comment]
@zalokhan did you update this? I think it needs a comment describing what you're doing in CorfuRuntimeTest (simulating failed connections).

**Comment by zalokhan** [comment]
@no2chem Added the comment.

**Pull request merged**